### PR TITLE
021 genderize first name

### PR DESCRIPTION
**Pull request by b264**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -263,6 +263,14 @@ FactoryHelper::Name.prefix #=> "Mr."
 FactoryHelper::Name.suffix #=> "IV"
 
 FactoryHelper::Name.title #=> "Legacy Creative Director"
+
+# NOTE NOTE
+# For the 'US only' methods below, first you must do the following:
+FactoryHelper::Config.locale = 'en-US'
+
+# US only
+FactoryHelper::Name.girls_name #=> "Marissa"
```
Comment (NikitaAvvakumov) [inline README.md]: What about scoping this under `en`? Even though it has US data, it should be broadly applicable to all English-speaking locales. In the future, if we have data for UK, CA, AUS etc., we can narrow the scopes again.

File: README.md [hunk 2]
```diff
@@ -308,7 +316,7 @@ FactoryHelper::PhoneNumber.phone_number #=> "397.693.1309"
 
 FactoryHelper::PhoneNumber.cell_phone #=> "(186)285-7925"
 
-# NOTE NOTE NOTE NOTE
+# NOTE NOTE
```
Comment (NikitaAvvakumov) [inline README.md]: Any more US-only examples to follow?

File: README.md [hunk 3]
```diff
@@ -263,6 +263,14 @@ FactoryHelper::Name.prefix #=> "Mr."
...
 
 FactoryHelper::Name.title #=> "Legacy Creative Director"
+
+# NOTE NOTE
+# For the 'US only' methods below, first you must do the following:
+FactoryHelper::Config.locale = 'en-US'
+
+# US only
+FactoryHelper::Name.girls_name #=> "Marissa"
+FactoryHelper::Name.boys_name #=> "Brian"
```
Comment (NikitaAvvakumov) [inline README.md]: `male_name`/`female_name`?

File: lib/locales/en-US.yml [hunk 4]
```diff
@@ -81,3 +81,6 @@ en-US:
...
...ana", "Vivien", "Vivienne", "Voncile", "Vonda", "Vonnie", "Wai", "Walter", "Waltraud", "Wan", "Wanda", "Waneta", "Wanetta", "Wanita", "Wava", "Wei", "Wen", "Wendi", "Wendie", "Wendolyn", "Wendy", "Wenona", "Wesley", "Whitley", "Whitney", "Wilda", "Wilhelmina", "Wilhemina", "Willa", "Willena", "Willene", "Willetta", "Willette", "Willia", "William", "Willie", "Williemae", "Willodean", "Willow", "Wilma", "Windy", "Winifred", "Winnie", "Winnifred", "Winona", "Winter", "Wonda", "Wynell", "Wynona", "Xenia", "Xiao", "Xiomara", "Xochitl", "Xuan", "Yadira", "Yaeko", "Yael", "Yahaira", "Yajaira", "Yan", "Yang", "Yanira", "Yasmin", "Yasmine", "Yasuko", "Yee", "Yelena", "Yen", "Yer", "Yesenia", "Yessenia", "Yetta", "Yevette", "Yi", "Ying", "Yoko", "Yolanda", "Yolande", "Yolando", "Yolonda", "Yon", "Yong", "Yoshie", "Yoshiko", "Youlanda", "Young", "Yu", "Yuette", "Yuk", "Yuki", "Yukiko", "Yuko", "Yulanda", "Yun", "Yung", "Yuonne", "Yuri", "Yuriko", "Yvette", "Yvone", "Yvonne", "Zada", "Zaida", "Zana", "Zandra", "Zelda", "Zella", "Zelma", "Zena", "Zenaida", "Zenia", "Zenobia", "Zetta", "Zina", "Zita", "Zoe", "Zofia", "Zoila", "Zola", "Zona", "Zonia", "Zora", "Zoraida", "Zula", "Zulema", "Zulma"]
```
Comment (NikitaAvvakumov) [inline lib/locales/en-US.yml]: As above

File: test/test_factory_helper_name.rb [hunk 5]
```diff
@@ -3,18 +3,38 @@
...
+      assert @tester.girls_name.match(/^([A-Z])(\w)*$/)
+    end
+  end
+
+  def test_boys_name_en_US
+    FactoryHelper::Config.locale= 'en-US'
+    300.times do
+      assert @tester.boys_name.match(/^([A-Z])(\w)*$/)
+    end
+  end
```
Comment (NikitaAvvakumov) [inline test/test_factory_helper_name.rb:52]: So, based on our discussion, what about separating test of values' validity from test of ability to access said values:

``` ruby
en_us_locale = Psych.load(File.read('lib/locales/en-US.yml'))
boys_names = en_us_locale['en-US']['factory_helper']['name']['boys_name']
assert_equal true, boys_names.map { |n| n.match(/\A[A-Z][\w']+/) }.all? # apostrophe to match "D'Angelo" in `en` locale, might need to adjust further for the `en-US` names

assert @tester.boys_name.match(/\[A-Z][\w']+$/) # 1 time only to test retrieval
# or even better
assert_includes boys_names, @tester.boys_name
```

File: test/test_factory_helper_name.rb [hunk 6]
```diff
@@ -3,18 +3,38 @@
...
-  
+
   def test_suffix
     assert @tester.suffix.match(/[A-Z][a-z]*\.?/)
   end
+
+  def test_girls_name_en_US
+    FactoryHelper::Config.locale= 'en-US'
+    1200.times do
+      assert @tester.girls_name.match(/^([A-Z])(\w)*$/)
```
Comment (NikitaAvvakumov) [inline test/test_factory_helper_name.rb]: Parens in regex for clarity? Don't need otherwise

**Comment by b264** [comment]
updated

File: lib/locales/en.yml [hunk 7]
```diff
@@ -112,10 +113,12 @@ en:
...
..."Tad", "Tanner", "Taylor", "Ted", "Teddy", "Teodoro", "Terence", "Terrance", "Terrell", "Terrence", "Terry", "Thad", "Thaddeus", "Thanh", "Theo", "Theodore", "Theron", "Thomas", "Thurman", "Tim", "Timmy", "Timothy", "Titus", "Tobias", "Toby", "Tod", "Todd", "Tom", "Tomas", "Tommie", "Tommy", "Toney", "Tony", "Tory", "Tracey", "Tracy", "Travis", "Trent", "Trenton", "Trevor", "Trey", "Trinidad", "Tristan", "Troy", "Truman", "Tuan", "Ty", "Tyler", "Tyree", "Tyrell", "Tyron", "Tyrone", "Tyson", "Ulysses", "Val", "Valentin", "Valentine", "Van", "Vance", "Vaughn", "Vern", "Vernon", "Vicente", "Victor", "Vince", "Vincent", "Vincenzo", "Virgil", "Virgilio", "Vito", "Von", "Wade", "Waldo", "Walker", "Wallace", "Wally", "Walter", "Walton", "Ward", "Warner", "Warren", "Waylon", "Wayne", "Weldon", "Wendell", "Werner", "Wes", "Wesley", "Weston", "Whitney", "Wilber", "Wilbert", "Wilbur", "Wilburn", "Wiley", "Wilford", "Wilfred", "Wilfredo", "Will", "Willard", "William", "Williams", "Willian", "Willie", "Willis", "Willy", "Wilmer", "Wilson", "Wilton", "Winford", "Winfred", "Winston", "Wm", "Woodrow", "Wyatt", "Xavier", "Yong", "Young", "Zachariah", "Zachary", "Zachery", "Zack", "Zackary", "Zane"]
```
Comment (NikitaAvvakumov) [inline lib/locales/en.yml]: Quotes should be unnecessary

File: lib/factory-helper/name.rb [hunk 8]
```diff
@@ -8,14 +8,35 @@ def name
         parse('name.name')
       end
 
-      def first_name; fetch('name.first_name'); end
-      def last_name;  fetch('name.last_name'); end
-      def prefix;     fetch('name.prefix'); end
-      def suffix;     fetch('name.suffix'); end
+      def first_name
+        fetch('name.first_name')
```
Comment (NikitaAvvakumov) [inline lib/factory-helper/name.rb:12]: Would it make sense to randomly fetch `male_name` or `female_name`, with a fallback to `first_name` for locales that don't define those?

Comment (b264) [inline lib/factory-helper/name.rb:12]: The translate method as defined will not allow this.  It will default the first lookup to `en` if
 undefined

**Comment by NikitaAvvakumov** [comment]
:shipit:

That said, I have two points to raise:
1. Is there a need to include quite that many female names? 4 times more than male names and (rough estimate) 35% more than the `first_names` we had to begin with.
1. While the test here is consistent with the legacy ones, I'm having a hard time understanding the advantage of this approach vs. the one in one of my comments. The current test relies on repetitive exercise of the suite to catch any invalid values (that is, those that don't match a very generous regex). And it still leaves open the possibility that the method implementation is screwed up and e.g. instead of male names retrieves female names, or street names, or business titles, etc.
   By contrast, what we can do is: first, iterate over the entire array and verify that values match the regex (if this is even important); second, make sure that the method retrieves values from said array and not another.

**Pull request merged**